ibm/resource_ibm_cos_bucket.go

```diff
@@ -935,41 +1003,69 @@ func resourceIBMCOSBucketDelete(d *schema.ResourceData, meta interface{}) error
 	s3Sess := session.Must(session.NewSession())
 	s3Client := s3.New(s3Sess, s3Conf)
 
-	if delbucket, ok := d.GetOk("force_delete"); ok {
-		if delbucket.(bool) {
+	delete := &s3.DeleteBucketInput{
+		Bucket: aws.String(bucketName),
+	}
+	_, err = s3Client.DeleteBucket(delete)
+	if err != nil {
+		return err
+	}
+	d.SetId("")
+	if err != nil && strings.Contains(err.Error(), "BucketNotEmpty") {
+		log.Println("bucket is not empty")
+		if delbucket, ok := d.GetOk("force_delete"); ok {
+			if delbucket.(bool) {
+				// Use a S3 service client that can handle multiple slashes in URIs.
+				// While aws_s3_bucket_object resources cannot create these object
+				// keys, other AWS services and applications using the S3 Bucket can.
 
-			// List objects within a bucket
-			resp, err := s3Client.ListObjects(&s3.ListObjectsInput{Bucket: aws.String(bucketName)})
-			if err != nil {
-				return fmt.Errorf("Unable to list items in bucket %s, %v", bucketName, err)
-			}
-			for _, item := range resp.Contents {
-				// Delete object within the bucket
-				_, err = s3Client.DeleteObject(&s3.DeleteObjectInput{Bucket: aws.String(bucketName), Key: aws.String(*item.Key)})
+				// bucket may have things delete them
+				log.Printf("[DEBUG] COS Bucket attempting to forceDelete %+v", err)
 
-				if err != nil {
-					return fmt.Errorf("Unable to delete object %s from bucket %s, %v", *item.Key, bucketName, err)
-				}
+				// Delete everything including locked objects.
+				// Don't ignore any object errors or we could recurse infinitely.
+				err = deleteAllCOSObjectVersions(s3Client, d.Id(), "", false, false)
```

Comment:
d.ID() is combination of multiple attributes but here we need to send the bucket name